tree/30_generic_methods/log_rudder.cf

```diff
@@ -0,0 +1,282 @@
+      # provided expected reports file has changed). If so, expand the variables
+      # in it into a temporary file.
+    !logger_rudder_temp_resfile_repaired::
+      "${expected_reports_temp}"
+        create        => "true",
+        edit_template => "${expected_reports_source}",
+        classes       => classes_generic("logger_rudder_temp_resfile");
+
+      # 3/ Write the final expected reports, expanded. First, empty it, then fill it
+      # (can't use edit_default empty, as we are iterating over a list)
+    logger_rudder_temp_resfile_repaired.!logger_rudder_final_resfile_repaired.keys_defined::
+      "${expected_reports_file}"
+        create => "true",
+        edit_line => delete_lines_matching(".*");
+
+    logger_rudder_temp_resfile_repaired.!logger_rudder_final_resfile_repaired.keys_defined::
+      "${expected_reports_file}"
+        create    => "true",
+        edit_line => insert_lines("${reports[${keys}][0]};;${reports[${keys}][canon]};;${reports[${keys}][2]};;${reports[${keys}][3]};;${reports[${keys}][4]};;${reports[${keys}][5]};;${reports[${keys}][6]}"),
+        classes   => classes_generic("logger_rudder_final_resfile");
+
+  methods:
+    # no need to avoid normal ordering since this class is set when report is ready
+    # Old reporting method, and data is ready
+    !report_v2.report_data_read.!no_old_class_prefix::
+
+      # 4/ Array is ready, reporting time !!!
+      "any" usebundle => _rudder_common_reports_generic("${report_data[1]}", "${c_class_prefix}", "${report_data[3]}", "${report_data[4]}", "${report_data[5]}", "${message}"),
+        classes    => classes_generic("logger_rudder_${c_class_prefix}");
+
+    # We HAVE the data and we CAN dot it (see /4 abvove) -> New reporting method
+    report_v2::
+      "any" usebundle => _log_rudder_v2("${expected_reports_source}", "${message}", "${origin_class_prefix}", @{args});
+}
+
+# Bundle that is called only if we have v2 expected report and the promiser stack
+#
+# It produces new style reports
+# It will completly override log_rudder when we remove v1 
+# Since it uses a new logic and need to parse the original file directly it can be put appart to avoid messing with the original code
+#
+# There is one case it can conflict with the orignal one :
+#  - the expected report contains one line with a v2 style technique that matches current call
+#  - the expected report contains one line with an unrelated v1 style technique that matches the current call too
+# The answer is that you should re-run rudderify on all your techniques when you upgrade
+bundle agent _log_rudder_v2(expected_reports_source, message, class_prefix, args)
+{
+  vars:
+      # head(length-3) -> remove log call stack (_log + log_rudder + _log_rudder or _logger + _log + log_rudder + _log_rudder or logger_rudder + log_rudder + _log_rudder)
+      "promiser_count"  int => length("this.callers_promisers");
+      "nologger_len" string => eval("${promiser_count}-3", "math", "infix");
```

Comment:
Yes, but we will require users to update away from 3.5 agents (Rudder 2.10) before upgrading.